### PR TITLE
Add CFUN implementation

### PR DESCRIPTION
**Pull request by Oddin60F**
This PR implements the CFUN algorithm for discovering constant conditional functional dependencies within the dataset.

The implementation follows the approach detailed in the paper:
>  "Discovering (frequent) Constant Conditional Functional Dependencies" by Thierno Diallo, Noël Novelli, and Jean-Marc Petit.

Added example Python script demonstrating how to use the CFUN algorithm from the Python API.

### CODE REVIEW

File: src/python_bindings/cfd/bind_cfd.cpp [hunk 1]
```diff
@@ -82,10 +96,7 @@ void BindCfd(py::module_& main_module) {
...
                     }));
 
-    py::class_<RawCFD>(cfd_module, "CFD")
-            .def(py::init<>())
-            .def(py::init<RawCFD::RawItems, RawCFD::RawItem>(), py::arg("lhs"), py::arg("rhs"))
-            .def(py::init(&CreateRawCFDFromTuples), py::arg("lhs"), py::arg("rhs"))
+    py::class_<RawCFD>(fd_first_module, "CFD")
             .def("__str__", &RawCFD::ToString)
             .def_property_readonly("lhs", &RawCFD::GetLhs)
             .def_property_readonly("rhs", &RawCFD::GetRhs)
```
Comment (VinogradovDmitri) [inline src/python_bindings/cfd/bind_cfd.cpp:102]: [Here](https://github.com/Desbordante/desbordante-core/wiki/Development%E2%80%90Features#%D1%87%D1%82%D0%BE-%D0%B7%D0%BD%D0%B0%D1%87%D0%B8%D1%82-%D1%80%D0%B0%D0%B7%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0%D1%82%D1%8C-%D0%B1%D0%B0%D0%B9%D0%BD%D0%B4%D0%B8%D0%BD%D0%B3%D0%B8)
```suggestion
            .def("__str__", &RawCFD::ToString)
            .def("__repr__", &CCFD::ToString)
            .def_property_readonly("lhs", &RawCFD::GetLhs)
            .def_property_readonly("rhs", &RawCFD::GetRhs)
```

File: src/python_bindings/cfd/bind_cfd.cpp [hunk 2]
```diff
@@ -103,8 +114,107 @@ void BindCfd(py::module_& main_module) {
...
+void BindCfun(pybind11::module_& cfd_module) {
+    using namespace algos::cfd::cfun;
+    auto cfun_module = python_bindings::BindPrimitiveNoBase<CFUN>(cfd_module, "CFUN")
+                               .def("get_cfds", &CFUN::GetCFDList);
+
+    py::class_<CCFD::Condition>(cfun_module, "Condition")
+            .def_property_readonly("pattern", &CCFD::Condition::GetPattern)
+            .def_property_readonly("support", &CCFD::Condition::GetSupport)
+            .def(py::self == py::self)
+            .def("__hash__",
```
Comment (VinogradovDmitri) [inline src/python_bindings/cfd/bind_cfd.cpp:128]: [Here](https://github.com/Desbordante/desbordante-core/wiki/Development%E2%80%90Features#%D1%87%D1%82%D0%BE-%D0%B7%D0%BD%D0%B0%D1%87%D0%B8%D1%82-%D1%80%D0%B0%D0%B7%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0%D1%82%D1%8C-%D0%B1%D0%B0%D0%B9%D0%BD%D0%B4%D0%B8%D0%BD%D0%B3%D0%B8) `Condition` gets neither `__str__` nor `__repr__`
(only `pattern`/`support` properties, `==`, `__hash__`, pickle), so printing a
tableau row shows `<desbordante.cfd.algorithms.CFUN.Condition object at 0x…>`.
`CCFD` defines `__str__` but no `__repr__`.

```suggestion
    py::class_<CCFD::Condition>(cfun_module, "Condition")
            .def(py::init([](std::vector<std::string> pattern, size_t support) {
                     return CCFD::Condition(std::move(pattern), support);
                 }),
                 py::arg("pattern"), py::arg("support"))
            .def_property_readonly("pattern", &CCFD::Condition::GetPattern)
            .def_property_readonly("support", &CCFD::Condition::GetSupport)
            .def("__repr__",
                 [](CCFD::Condition const& condition) {
                     std::string repr = "Condition(pattern=[";
                     for (auto const& value : condition.GetPattern()) {
                         if (repr.back() != '[') {
                             repr += ", ";
                         }
                         repr += "'" + value + "'";
                     }
                     repr += "], support=";
                     repr += std::to_string(condition.GetSupport());
                     repr += ")";
                     return repr;
                 })
            .def(py::self == py::self)
            .def("__hash__",
```

File: src/core/algorithms/cfd/cfun/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+set(NAME cfd.cfun)
+desbordante_add_lib(NAME)
+target_sources(${NAME} PRIVATE cfun.cpp quadruple.cpp attribute_utils.cpp)
+target_link_libraries(
+    ${NAME}
+    PRIVATE ${DESBORDANTE_PREFIX}::cfd::util ${DESBORDANTE_PREFIX}::cfd::model
+            ${DESBORDANTE_PREFIX}::algos ${DESBORDANTE_PREFIX}::fd spdlog::spdlog_header_only
+            Boost::headers
```
Comment (VinogradovDmitri) [inline src/core/algorithms/cfd/cfun/CMakeLists.txt:8]: [Here](https://github.com/Desbordante/desbordante-core/wiki/Development:-target%E2%80%90based-CMake-quick-start-guide#target-based-cmake-quick-start-guide)

```suggestion
    PUBLIC ${DESBORDANTE_PREFIX}::algos ${DESBORDANTE_PREFIX}::config
           ${DESBORDANTE_PREFIX}::cfd::model ${DESBORDANTE_PREFIX}::fd Boost::headers
    PRIVATE ${DESBORDANTE_PREFIX}::util spdlog::spdlog_header_only
```

File: src/core/algorithms/cfd/cfun/quadruple.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
+#include "core/algorithms/cfd/cfun/quadruple.h"
+
+#include <cstddef>
+#include <unordered_map>
+#include <unordered_set>
```
Comment (VinogradovDmitri) [inline src/core/algorithms/cfd/cfun/quadruple.cpp:5]: Delete all unused includes, for example this

File: src/core/algorithms/cfd/cfun/cfun.cpp [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+void CFUN::MakeExecuteOptsAvailable() {
+    using namespace config::names;
+    MakeOptionsAvailable({kCfdMinimumSupport});
+}
+
+void CFUN::LoadDataInternal() {
+    cfd_relation_ = CFDRelationData::CreateFrom(*input_table_);
+    if (cfd_relation_->GetColumnData().empty()) {
```
Comment (VinogradovDmitri) [inline src/core/algorithms/cfd/cfun/cfun.cpp:35]: The empty-dataset guard only rejects tables with zero *columns* (`GetColumnData().empty()`). A header-only CSV loads successfully; then `ExecuteInternal` reaches `CheckCFD`, which reads `target_col_ec[cluster.indices[0]]` on the zero-level candidate whose single cluster has **empty** `indices` — an out-of-bounds read (UB)

```suggestion
    if (cfd_relation_->GetColumnData().empty() || cfd_relation_->GetNumRows() == 0) {
```

**Review by VinogradovDmitri — CHANGES_REQUESTED** [review]
So far this

File: src/core/algorithms/cfd/cfun/cfun.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    std::list<CCFD> const& GetCFDList() const noexcept {
+        return cfd_list_;
+    }
+
+    void MakeExecuteOptsAvailable() override;
+    void RegisterOptions();
+    void ExecuteInternal() override;
+    void LoadDataInternal() override;
+
+    void ResetState() final {}
```
Comment (VinogradovDmitri) [inline src/core/algorithms/cfd/cfun/cfun.h:54]:
```suggestion
    void ResetState() final;
```

File: src/core/algorithms/cfd/cfun/cfun.cpp [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+            if (candidate_ij.IsEmptySets()) {
+                continue;
+            }
+
+            current.push_back(std::move(candidate_ij));
+        }
+    }
+}
+
```
Comment (VinogradovDmitri) [inline src/core/algorithms/cfd/cfun/cfun.cpp:238]: the `Algorithm` contract states `ResetState` exists to "Clear the necessary fields for Execute to run repeatedly with different configuration parameters on the same dataset"
```
void CFUN::ResetState() {
    cfd_list_.clear();
}
```

File: src/core/algorithms/cfd/cfun/cfun.cpp [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+CFUN::CFUN() : Algorithm() {
+    using namespace config::names;
+    RegisterOptions();
+    MakeOptionsAvailable({kTable});
+}
+
+void CFUN::RegisterOptions() {
+    DESBORDANTE_OPTION_USING;
+    RegisterOption(config::kTableOpt(&input_table_));
+    RegisterOption(Option{&min_support_, kCfdMinimumSupport, kDCfdMinimumSupport, 0u});
```
Comment (VinogradovDmitri) [inline src/core/algorithms/cfd/cfun/cfun.cpp:25]: CFUN registers the option with default `0u` and no
value check: `RegisterOption(Option{&min_support_, kCfdMinimumSupport,
kDCfdMinimumSupport, 0u});`. The new example tells users the opposite:
"Note that cfd_minsup must be greater than 0."

```suggestion
    RegisterOption(config::kTableOpt(&input_table_));

    auto check_supp = [](unsigned int value) {
        if (value == 0) {
            throw config::ConfigurationError("Minimum support must be greater than 0.");
        }
    };
    RegisterOption(
            Option{&min_support_, kCfdMinimumSupport, kDCfdMinimumSupport, 1u}.SetValueCheck(
                    std::move(check_supp)));
```

File: src/core/algorithms/cfd/cfun/cfun.h (hunk 6 above)
Comment (VinogradovDmitri) [inline src/core/algorithms/cfd/cfun/cfun.h:54]: the base class declares `MakeExecuteOptsAvailable`, `ExecuteInternal`, `LoadDataInternal`, etc. `protected`/`private` (algorithm.h:37-38 private virtual,
algorithm.h:67 protected virtual

```suggestion
private:
    void MakeExecuteOptsAvailable() override;
    void RegisterOptions();
    void ExecuteInternal() override;
    void LoadDataInternal() override;

    void ResetState() final {}
```

File: src/tests/unit/test_cfun.cpp
Comment (VinogradovDmitri) [inline src/tests/unit/test_cfun.cpp:1]: add edge cases:
empty input, single row/column, all-duplicate rows, all-null column, maximum-cardinality columns

Also, snapshot does not match actual example output

File: src/core/algorithms/cfd/cfun/cfun.cpp [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+                break;
+            }
+
+            if (!attr_to_candidate.AllSubsetsContains(candidate_i.GetAttributes() |
+                                                      candidate_j.GetAttributes())) {
+                continue;
+            }
+
+            auto candidate_ij =
+                    candidate_i.Intersect(candidate_j, cfd_relation_->GetNumRows(), min_support_);
```
Comment (VinogradovDmitri) [inline src/core/algorithms/cfd/cfun/cfun.cpp:228]: every candidate-pair intersection zero-initializes and discards `probing`, `assoc`, `counts`, `elem_result` -- each `(num_rows+1)×4 B ≈ 32 KB` on mushroom (~130 KB churn per pair). heaptrack attributes 136,775 allocation calls / 83.30 MB peak to these `reserve`s, and Intersect itself takes 25.41% of cycles.
So we should reuse instead of reconstructing.

```suggestion
    auto attr_to_candidate = AttributeIndex(last);
    IntersectWorkspace workspace;

    for (size_t i = 0; i < last.size() - 1; i++) {
        for (size_t j = i + 1; j < last.size(); j++) {
            auto const& candidate_i = last[i];
            auto const& candidate_j = last[j];

            if (!candidate_i.HasSameBegin(candidate_j)) {
                break;
            }

            if (!attr_to_candidate.AllSubsetsContains(candidate_i.GetAttributes() |
                                                      candidate_j.GetAttributes())) {
                continue;
            }

            auto candidate_ij =
                    candidate_i.Intersect(candidate_j, cfd_relation_->GetNumRows(), min_support_,
                                                           workspace);
```

**Review by VinogradovDmitri — CHANGES_REQUESTED** [review]
